jbmc/src/java_bytecode/java_types.h

```diff
@@ -268,6 +298,7 @@ inline const java_method_typet &to_java_method_type(const typet &type)
 inline java_method_typet &to_java_method_type(typet &type)
 {
   PRECONDITION(type.id() == ID_code);
+  type.set(ID_C_java_method_type, true);
```

Comment:
:no_entry_sign: Do this in the constructor only and check it here. That will catch any cases where something is expecting a java_method_typet but actually got a regular old `code_typet`

Reply:
Best: introduce:
```
template <>
inline bool can_cast_type<java_method_typet>(const typet &type)
{
  return type.id() == ID_code && type.get_bool(ID_C_java_method_type);
}
```

Above the two `to_java_method_type` and use it in the pre-condition of both.